schematools/validation.py

```diff
@@ -0,0 +1,149 @@
+"""Semantic JSON schema validation.
+
+This package provides a simple :class:`Validator` class
+that allows us to perform semantic validation
+on a given JSON schema.
+
+Semantic validation is different from structural validation,
+as performed by the :module:`jsonschema`,
+in that it takes into account the implied logic behind a JSON schema.
+For instance,
+we might indicate that what identifies a table
+is a set of properties (columns).
+In Amsterdam Schema this is done using the ``identifier`` property
+that can take an array of property names.
+Structural validation cannot verify that the properties specified in that array
+actually do exist in the schema.
+Semantic validation can and should.
+
+Specific semantic validations are created by subclassing :class:`Validator`
+and overriding the :meth:`Validator.validate` method.
+Simply by means of subclassing,
+the new validator is automatically registered.
+Registered validator classes will all be run
+when :meth:`Validator.run_all()` is invoked.
+
+.. note::
+
+   For the registration to work
+   all :class:`Validator` subclasses need to be parsed by the Python interpreter.
+   This can be achieved by importing the module they reside in.
+
+Example:
+    The following will run all registered validators
+    on a dataset ``dataset``.
+    Any validation errors are printed to ``stdout``::
+
+        dataset = _get_dataset_schema(meta_schema_url, schema_location)
+        validator = Validator(dataset=dataset)
+        for error in validator.run_all():
+            print(error)
+
+"""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import ClassVar, Iterator, List, Type, final
+
+from schematools import MAX_TABLE_NAME_LENGTH
+from schematools.types import DatasetSchema
+from schematools.utils import to_snake_case
+
+
+@dataclass(frozen=True)
+class ValidationError:
+    """Capture validation errors."""
+
+    validator_name: str
+    message: str
+
+    def __str__(self) -> str:
+        return f"[{self.validator_name}] {self.message}"
+
+
+class ValidationException(Exception):
+    """Raised when validation fails to execute.
+
+    .. note::
+
+       This is not for validation errors. See :class:`ValidationError` instead.
+    """
+
+    message: str
+
+    def __init__(self, message: str) -> None:
+        super().__init__(*message)
+        self.message = message
+
+
+class Validator:
+    """Base class for validators.
+
+    Not only is this a base class for validators,
+    it is also used for running all registered validators.
+    See Also: :meth:`run_all`
+
+    Registration is a side-effect of overriding this base class.
+
+    """
+
+    _registry: ClassVar[List[Type[Validator]]] = []
+    dataset: DatasetSchema
+
+    @classmethod
+    def __init_subclass__(cls: Type[Validator]) -> None:
```

Comment:
Nice approach, this registration!

Reply:
Yeah, I saw a former co-worker use this once and thought "Does this really do what it seems to be doing?" Turns out it does: https://docs.python.org/3.8/reference/datamodel.html?highlight=__init_subclass__#object.__init_subclass__

New since 3.6. There's so much stuff in Python. I often feel I don't know half of it.